### PR TITLE
Deprecate `PhysicalSortRequirement::from_sort_exprs` and `PhysicalSortRequirement::to_sort_exprs`

### PR DESCRIPTION
**Pull request by alamb**
## Which issue does this PR close?

Part of https://github.com/apache/datafusion/issues/12591

## Rationale for this change

Before `LexOrdering` and `LexRequirement` were actual structs (thanks to @ngli-me  and @berkaysynnada  on https://github.com/apache/datafusion/pull/13146!!!), we couldn't put the conversion between `LexOrdering` and `LexRequirement` on to the actual structures and thus the conversion is done via free functions on `PhysicalSortRequirements`

This is confusing for several reasons:
- The conversion is on another struct (so when Rust programmers go to look for a `From` conversion they aren't there)
- The conversions hide a clone when ideomatic Rust mostly prefers explict cloning when it is occuring

This structure bothered me for a long time, so it is time to scratch an itch now that @ngli-me  started the ball rolling

## What changes are included in this PR?

1. Deprecate `PhysicalSortRequirement::from_sort_exprs` and `PhysicalSortRequirement::to_sort_exprs`
2. Add `LexOrdering::from_lex_requirement` and `LexOrdering::from_lex_ordering` and `From` impls
3. Update code

## Are these changes tested?

By existint CI
## Are there any user-facing changes?

SOme methods are deprecated with hints about what to use instead

### CODE REVIEW

File: datafusion/core/src/physical_optimizer/enforce_distribution.rs [hunk 1]
```diff
@@ -1492,9 +1491,7 @@ pub(crate) mod tests {
             if self.expr.is_empty() {
                 vec![None]
             } else {
-                vec![Some(PhysicalSortRequirement::from_sort_exprs(
-                    self.expr.iter(),
-                ))]
+                vec![Some(LexRequirement::from(self.expr.clone()))]
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforce_distribution.rs:1494]: This is a pretty good example of the potential improvement -- it is now clear this function is returning a `LexRequirement` and that it clones the underlying `LexOrder` structure

File: datafusion/core/src/physical_optimizer/enforce_sorting.rs [hunk 2]
```diff
@@ -221,7 +221,7 @@ fn replace_with_partial_sort(
         // here we're trying to find the common prefix for sorted columns that is required for the
         // sort and already satisfied by the given ordering
         let child_eq_properties = child.equivalence_properties();
-        let sort_req = PhysicalSortRequirement::from_sort_exprs(sort_plan.expr());
+        let sort_req = LexRequirement::from(sort_plan.expr().clone());
```
Comment (alamb) [inline datafusion/core/src/physical_optimizer/enforce_sorting.rs:224]: Again, there is no more cloning going on than previously, just now the clone is explicit

**Review by berkaysynnada — APPROVED** [review]
Looks great, the code is clearly improving. Thank you @alamb.

**Comment by alamb** [comment]
Thank you for the review @berkaysynnada

**Pull request merged**